File: Trees/Python/Root to Leaf Paths.py
# Problem link: https://www.geeksforgeeks.org/problems/root-to-leaf-paths/1
# Time: O(n)
# Space: O(height of the tree)
class Solution:
    def Paths(self, root : Optional['Node']) -> List[List[int]]:
        result, lst = [], []
        
        def dfs(node):
            if node and not node.left and not node.right:
                lst.append(node.data)
                result.append(lst[:])
                lst.pop()
            elif node:
                lst.append(node.data)
                dfs(node.left)
                dfs(node.right)
                lst.pop()
        
        dfs(root)
        
        return result

File: Trees/Java/Root to Leaf Paths.java
// Problem link: https://www.geeksforgeeks.org/problems/root-to-leaf-paths/1
// Time: O(n)
// Space: O(height of the tree)
class Solution {
    public static ArrayList<ArrayList<Integer>> Paths(Node root) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<>();
        ArrayList<Integer> lst = new ArrayList<>();
        
        dfs(root, result, lst);
        
        return result;
    }
    
    private static void dfs(Node node, ArrayList<ArrayList<Integer>> result, ArrayList<Integer> lst) {
        if (node != null && node.left == null && node.right == null) {
            lst.add(node.data);
            result.add(new ArrayList<>(lst));
            lst.remove(lst.size() - 1);
        } else if (node != null) {
            lst.add(node.data);
            dfs(node.left, result, lst);
            dfs(node.right, result, lst);
            lst.remove(lst.size() - 1);
        }
    }
}
